### PR TITLE
merykitty's attempt

### PR DESCRIPTION
**Pull request by merykitty**
My attempt at this challenge, `test.sh merykitty` passed.

Please kindly review, thanks very much.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Oh, wow, this looks very promising. 7.379s on the eval env, fastest so far! Summoning the current top of the pack @royvanrijn, @ebarlas, @spullara, for feedback too!

**Comment by merykitty** [comment]
I just fixed an edge case of overlapping batches, that does not affect the overall performance, though, but you may want to run again just to be sure. Running this on my machine shows that the throughput is bottlenecked by the performance of the `mmap` call. I theorise it may be better to fetch the file concurrently with the processing in a manner similar to prefetching but have not been able to materialise the idea.

image

**Comment by merykitty** [comment]
It turns out that just prefetching the file manually helps lower the time by 20% on my machine, now it takes around 2 seconds, for comparison the baseline takes around 106 seconds and just reading the whole file without doing anything takes around 0.75 seconds.

Thanks.

File: src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        }
+        point = negative ? -point : point;
+        aggr.min = Math.min(point, aggr.min);
+        aggr.max = Math.max(point, aggr.max);
+        aggr.sum += point;
+        aggr.count++;
+        return offset;
+    }
+
+    private static long iterate(PoorManMap aggrMap, MemorySegment data, long offset) {
```
Comment (gunnarmorling) [inline src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java:207]: Could you provide a description of how that iteration loop works? I roughly get it, but it's quite dense code and having a high-level description would be great.

File: src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        long word = data.get(JAVA_LONG_LT, offset);
+        // This can be 12, 20, 28
+        int decimalSepPos = Long.numberOfTrailingZeros(~word & 0x10101000);
+        int shift = 28 - decimalSepPos;
+        // signed is -1 if negative, 0 otherwise
+        long signed = (~word << 59) >> 63;
+        long designMask = ~(signed & 0xFF);
+        long digits = ((word & designMask) << shift) & 0x0F000F0F00L;
+
+        // Now digits is in the form 0xUU00TTHH00
```
Comment (gunnarmorling) [inline src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java:164]: What is TT, HH, etc? As for the main loop, could you provide an overview of how this thing works?

**Review by gunnarmorling — COMMENTED** [review]
Very nice! This comes in at 7.6, i.e. first place currently! I've added some questions in line to further my (and others) understanding. It looks alright to me, but getting some more intuition would be great. Thanks a lot!

**Comment by merykitty** [comment]
After trying again without prefetching I am also able to reach 2s, which is weird.

Thanks for your reviews, hope I have addressed those.

File: src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java (hunk 2 above)
Comment (merykitty) [reply src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java]: UU stands for the value of the units digit, TT is the tens digit and HH is the hundreds digit.

File: src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java (hunk 1 above)
Comment (merykitty) [reply src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java:217]: Done

File: src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        point = negative ? -point : point;
+        aggr.min = Math.min(point, aggr.min);
+        aggr.max = Math.max(point, aggr.max);
+        aggr.sum += point;
+        aggr.count++;
+        return offset;
+    }
+
+    private static long iterate(PoorManMap aggrMap, MemorySegment data, long offset) {
+        var line = ByteVector.fromMemorySegment(BYTE_SPECIES, data, offset, ByteOrder.nativeOrder());
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java:220]: No if we cannot find the delimiter which means the key is longer than the segment we fall back to slow path which do so in a more straightforward way,

[the remark this answers is not in the document]
Comment (gunnarmorling) [reply src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java:220]: Ah yes, got it. Thanks for adding all the comments!

**Comment by gunnarmorling** [comment]
00:07.620, current #1! Very nice, thanks for participating!

**Comment by DamienOReilly** [comment]
How does this one fare our with Graal?

**Pull request merged**